File: gymtracker-pro/frontend/src/utils/nutrition.ts
// Mifflin-St Jeor BMR calculation
export function calculateBMR(weightKg: number, heightCm: number, ageYears: number, sex: "M" | "F"): number {
  if (sex === "M") {
    return 10 * weightKg + 6.25 * heightCm - 5 * ageYears + 5;
  } else {
    return 10 * weightKg + 6.25 * heightCm - 5 * ageYears - 161;
  }
}

// Calculate TDEE based on activity level
export function calculateTDEE(bmr: number, activityLevel: "sedentary" | "light" | "moderate" | "active" | "very_active"): number {
  const multipliers: Record<typeof activityLevel, number> = {
    sedentary: 1.2,
    light: 1.375,
    moderate: 1.55,
    active: 1.725,
    very_active: 1.9,
  };
  return Math.round(bmr * multipliers[activityLevel]);
}

// Get calorie target based on goal
export function getCalorieTarget(tdee: number, goal: "CUTTING" | "BULKING" | "MAINTENANCE"): number {
  if (goal === "CUTTING") return Math.round(tdee * 0.8); // 20% deficit
  if (goal === "BULKING") return Math.round(tdee * 1.1); // 10% surplus
  return tdee;
}

// Get macro targets based on calories, weight, and goal
export function getMacroTargets(
  calories: number,
  weightKg: number
): { proteinG: number; carbsG: number; fatG: number } {
  const proteinG = Math.round(weightKg * 2.2); // 2.2g per kg
  const fatG = Math.round((calories * 0.25) / 9); // 25% of calories
  const carbsG = Math.round(((calories - proteinG * 4 - fatG * 9) / 4));

  return { proteinG, carbsG, fatG };
}

// Estimate 1RM using Epley formula
export function estimateOneRM(weightKg: number, reps: number): number {
  return Number((weightKg * (1 + reps / 30)).toFixed(1));
}

// Format date for display
export function formatDate(date: Date | string): string {
  if (typeof date === "string") date = new Date(date);
  return new Intl.DateTimeFormat("en-US", { month: "short", day: "numeric", year: "numeric" }).format(date);
}

// Format time for display
export function formatTime(date: Date | string): string {
  if (typeof date === "string") date = new Date(date);
  return new Intl.DateTimeFormat("en-US", { hour: "2-digit", minute: "2-digit" }).format(date);
}

// Get day name from date
export function getDayName(date: Date | string): string {
  if (typeof date === "string") date = new Date(date);
  return new Intl.DateTimeFormat("en-US", { weekday: "short" }).format(date);
}

// Get muscle group color
export function getMuscleGroupColor(muscleGroup: string): string {
  const colors: Record<string, string> = {
    CHEST: "bg-red-500",
    BACK: "bg-blue-500",
    SHOULDERS: "bg-purple-500",
    BICEPS: "bg-yellow-500",
    TRICEPS: "bg-orange-500",
    LEGS: "bg-green-500",
    GLUTES: "bg-pink-500",
    CORE: "bg-indigo-500",
    CARDIO: "bg-cyan-500",
    FULL_BODY: "bg-gray-500",
  };
  return colors[muscleGroup] || "bg-gray-400";
}

// Get meal type emoji
export function getMealTypeEmoji(mealType: string): string {
  const emojis: Record<string, string> = {
    BREAKFAST: "🍳",
    LUNCH: "🥗",
    DINNER: "🍖",
    SNACK: "🍎",
    PRE_WORKOUT: "⚡",
    POST_WORKOUT: "💪",
  };
  return emojis[mealType] || "🍽️";
}

File: backend_fastapi/app/services/nutrition.py
"""
Nutrition calculations and utilities.
"""


def calculate_bmr(weight_kg: float, height_cm: float, age: int, sex: str) -> float:
    """
    Calculate Basal Metabolic Rate using Mifflin-St Jeor formula.
    
    Args:
        weight_kg: Body weight in kg
        height_cm: Height in cm
        age: Age in years
        sex: "M" or "F"
    
    Returns:
        BMR in calories per day
    """
    if sex.upper() == "M":
        return 10 * weight_kg + 6.25 * height_cm - 5 * age + 5
    else:  # Female
        return 10 * weight_kg + 6.25 * height_cm - 5 * age - 161


def calculate_tdee(bmr: float, activity_level: float = 1.55) -> float:
    """
    Calculate Total Daily Energy Expenditure.
    
    Activity levels:
        1.2: Sedentary
        1.375: Lightly active
        1.55: Moderately active (3-5 days/week)
        1.725: Very active
        1.9: Extremely active
    """
    return bmr * activity_level


def get_calorie_target(tdee: float, goal: str) -> float:
    """
    Calculate daily calorie target based on goal.
    
    Args:
        tdee: Total daily energy expenditure
        goal: "CUTTING", "BULKING", or "MAINTENANCE"
    
    Returns:
        Daily calorie target
    """
    multipliers = {
        "CUTTING": 0.80,      # 20% deficit
        "BULKING": 1.10,      # 10% surplus
        "MAINTENANCE": 1.0
    }
    return tdee * multipliers.get(goal, 1.0)


def get_macro_targets(calories: float, weight_kg: float) -> dict:
    """
    Calculate macro targets (protein, carbs, fat).
    
    Strategy:
        - Protein: 2.2g per kg (muscle preservation)
        - Fat: 25% of calories
        - Carbs: Remaining calories
    """
    protein_g = weight_kg * 2.2
    fat_calories = calories * 0.25
    fat_g = fat_calories / 9  # 9 cal per gram
    carb_calories = calories - (protein_g * 4) - fat_calories
    carb_g = carb_calories / 4  # 4 cal per gram
    
    return {
        "proteinG": round(protein_g, 1),
        "fatG": round(fat_g, 1),
        "carbsG": round(carb_g, 1)
    }
